From: Benjamin Berg <bberg@redhat.com>
Date: Fri, 25 Feb 2022 10:09:17 +0100
Subject: input: Fix FD handling and let input stream close it

The open() call needs to check against the FD being <0 for error
detection. Also, add a missing close in one of the error paths and
change the code to let the input stream close the FD when it is not
needed anymore.

We still open() it manually just because we need to do an ioctl once.

(cherry picked from commit 72469426b737f1398651efc360f62a86e3b2633b)
---
 src/linux/up-input.c | 18 ++++++++----------
 1 file changed, 8 insertions(+), 10 deletions(-)

diff --git a/src/linux/up-input.c b/src/linux/up-input.c
index ccd42c7..db96a9e 100644
--- a/src/linux/up-input.c
+++ b/src/linux/up-input.c
@@ -37,7 +37,6 @@ struct _UpInput
 
 	guint			 watched_switch;
 	int			 last_switch_state;
-	int			 eventfp;
 	struct input_event	 event;
 	gsize			 offset;
 	GIOChannel		*channel;
@@ -192,6 +191,7 @@ up_input_coldplug (UpInput *input, GUdevDevice *d)
 	glong bitmask[NBITS(SW_MAX)];
 	gint num_bits;
 	GIOStatus status;
+	int eventfd;
 
 	/* get sysfs path */
 	native_path = up_input_get_device_sysfs_path (d);
@@ -243,23 +243,25 @@ up_input_coldplug (UpInput *input, GUdevDevice *d)
 	}
 
 	/* open device file */
-	input->eventfp = open (device_file, O_RDONLY | O_NONBLOCK);
-	if (input->eventfp <= 0) {
+	eventfd = open (device_file, O_RDONLY | O_NONBLOCK);
+	if (eventfd < 0) {
 		g_warning ("cannot open '%s': %s", device_file, strerror (errno));
 		ret = FALSE;
 		goto out;
 	}
 
 	/* get initial state */
-	if (ioctl (input->eventfp, EVIOCGSW(sizeof (bitmask)), bitmask) < 0) {
+	if (ioctl (eventfd, EVIOCGSW(sizeof (bitmask)), bitmask) < 0) {
 		g_warning ("ioctl EVIOCGSW on %s failed", native_path);
+		close(eventfd);
 		ret = FALSE;
 		goto out;
 	}
 
 	/* create channel */
-	g_debug ("watching %s (%i)", device_file, input->eventfp);
-	input->channel = g_io_channel_unix_new (input->eventfp);
+	g_debug ("watching %s (%i)", device_file, eventfd);
+	input->channel = g_io_channel_unix_new (eventfd);
+	g_io_channel_set_close_on_unref (input->channel, TRUE);
 
 	/* set binary encoding */
 	status = g_io_channel_set_encoding (input->channel, NULL, &error);
@@ -290,7 +292,6 @@ out:
 static void
 up_input_init (UpInput *input)
 {
-	input->eventfp = -1;
 	input->last_switch_state = -1;
 }
 
@@ -309,11 +310,8 @@ up_input_finalize (GObject *object)
 
 	if (input->channel) {
 		g_io_channel_shutdown (input->channel, FALSE, NULL);
-		input->eventfp = -1;
 		g_io_channel_unref (input->channel);
 	}
-	if (input->eventfp >= 0)
-		close (input->eventfp);
 	G_OBJECT_CLASS (up_input_parent_class)->finalize (object);
 }
 
